pkg/argo/fetch.go

```diff
@@ -55,48 +84,63 @@ func Fetch(req *pb.FetchRequest) (*pb.FetchResponse, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	pod, err := getPodByStepGroup(wf, stepGroupName)
+	stepCnt := len(wf.Spec.Templates[0].Steps)
+	stepIdx, err := getStepIdx(wf, stepGroupName)
 	if err != nil {
 		return nil, err
 	}
 
-	if isPodPending(pod) {
-		return newFetchResponse(req, false, []string{}), nil
-	}
-
-	logs, newLogOffset, err := getPodLogs(pod.Name, req.LogOffset)
+	pod, err := getPodByStepGroup(wf, stepGroupName)
 	if err != nil {
 		return nil, err
 	}
+	eof := false // true if finish fetching the workflow logs
+	logs := []string{}
+
+	// An example log content:
+	// Step [1/3] Execute Code:
+	// repl -e "SELECT * FROM iris.train"
+	// Step [1/3] Log view: http://localhost:8001/workflows/default/steps-bdpff
+	// Step [1/3] Status: Pending
+	// Step [1/3] Status: Running
+	// Step [1/3] Status: ...
+	// Step [1/3] Status: Succeed/Failed
+	// ..
+	newOffset := req.LogOffset
+	if req.LogOffset == "" {
+		// return the log view url for the first call of step
+		url, e := logViewURL(wf.ObjectMeta.Namespace, wf.ObjectMeta.Name, stepGroupName)
+		if e != nil {
+			return nil, e
+		}
+		// the 1-th container execute `argoexec wait` to wait the preiority step, so package the 2-th container's command code.
+		execCode := fmt.Sprintf("%s %s", strings.Join(pod.Spec.Containers[1].Command, " "), strings.Join(pod.Spec.Containers[1].Args, " "))
+		logs = append(logs, fmt.Sprintf("Step: [%d/%d] Execute Code: %s", stepIdx, stepCnt, execCode))
+		logs = append(logs, fmt.Sprintf("Step: [%d/%d] Log View: %s", stepIdx, stepCnt, url))
+	}
 
-	finishFetchingCurrentPod := false // true if finish fetching the current pod logs
-	eof := false                      // true if finish fetching the workflow logs
-
-	// finish fetching the current pod logs when:
-	// 1. the Pod is completed(succeed/failed/error).
-	// 2. no updated logs.
-	if req.LogOffset == newLogOffset && isPodCompleted(pod) {
-		finishFetchingCurrentPod = true
+	// output the refreshed status if the Pod Phase changed.
+	if req.LogOffset != string(pod.Status.Phase) {
+		logs = append(logs, fmt.Sprintf("Step: [%d/%d] Status: %s", stepIdx, stepCnt, pod.Status.Phase))
+		newOffset = string(pod.Status.Phase)
```

Comment:
Why the offset is set to `phase`?

Reply:
> Use the Pod phase as the offset to avoid output the large duplicated logs.

I add some comment in the code.

Reply:
Renamed  `logOffset` to `stepPhase` which better to understand.